tests/journald_tests.cpp

```diff
@@ -877,6 +879,184 @@ TEST_P(JournaldLoggerTest, ROOT_LogToJournald)
+        std::string("fluentbit+logrotate")));
+
+
+// Sets up a TCP to mock the fluentbit component and then loads the
+// ContainerLogger module and runs a task.
+TEST_P(FluentbitLoggerTest, ROOT_LogToFluentbit)
+{
+  // Create a master, agent, and framework.
+  Try<Owned<cluster::Master>> master = StartMaster();
+  ASSERT_SOME(master);
+
+  // We'll need access to these flags later.
+  mesos::internal::slave::Flags flags = CreateSlaveFlags();
+
+  // Use the journald container logger.
+  flags.container_logger = JOURNALD_LOGGER_NAME;
+
+  Fetcher fetcher(flags);
+
+  // We use an actual containerizer + executor since we want something to run.
+  Try<MesosContainerizer*> _containerizer =
+    MesosContainerizer::create(flags, false, &fetcher);
+
+  CHECK_SOME(_containerizer);
+  Owned<MesosContainerizer> containerizer(_containerizer.get());
+
+  Owned<MasterDetector> detector = master.get()->createDetector();
+
+  Try<Owned<cluster::Slave>> slave =
+    StartSlave(detector.get(), containerizer.get(), flags);
+  ASSERT_SOME(slave);
+
+  MockScheduler sched;
+  MesosSchedulerDriver driver(
+      &sched, DEFAULT_FRAMEWORK_INFO, master.get()->pid, DEFAULT_CREDENTIAL);
+
+  Future<FrameworkID> frameworkId;
+  EXPECT_CALL(sched, registered(&driver, _, _))
+    .WillOnce(FutureArg<1>(&frameworkId));
+
+  // Wait for an offer, and start a task.
+  Future<vector<Offer>> offers;
+  EXPECT_CALL(sched, resourceOffers(&driver, _))
+    .WillOnce(FutureArg<1>(&offers))
+    .WillRepeatedly(Return()); // Ignore subsequent offers.
+
+  driver.start();
+  AWAIT_READY(frameworkId);
+
+  AWAIT_READY(offers);
+  EXPECT_NE(0u, offers.get().size());
+
+  const std::string specialString = "some-super-unique-string";
+
+  TaskInfo task = createTask(offers.get()[0], "echo " + specialString);
+
+  // Change the destination of the logs based on the parameterized test.
+  Environment::Variable* variable =
+    task.mutable_command()->mutable_environment()->add_variables();
+  variable->set_name("CONTAINER_LOGGER_DESTINATION_TYPE");
+  variable->set_value(GetParam());
+
+  Future<TaskStatus> statusStarting;
+  Future<TaskStatus> statusRunning;
+  Future<TaskStatus> statusFinished;
+  EXPECT_CALL(sched, statusUpdate(&driver, _))
+    .WillOnce(FutureArg<1>(&statusStarting))
+    .WillOnce(FutureArg<1>(&statusRunning))
+    .WillOnce(FutureArg<1>(&statusFinished))
+    .WillRepeatedly(Return());       // Ignore subsequent updates.
+
+  driver.launchTasks(offers.get()[0].id(), {task});
+
+  AWAIT_READY(statusStarting);
+  EXPECT_EQ(TASK_STARTING, statusStarting.get().state());
+
+  AWAIT_READY(statusRunning);
+  EXPECT_EQ(TASK_RUNNING, statusRunning.get().state());
+
+  AWAIT_READY(statusFinished);
+  EXPECT_EQ(TASK_FINISHED, statusFinished.get().state());
+
+  driver.stop();
+  driver.join();
+
+  // We expect two connections to be made:
+  // one for stdout and one for stderr.
+  Future<network::inet::Socket> _client1 = fluentbit_server->accept();
```

Comment:
The clients can successfully connect as soon as this line: 
`Try<Nothing> listen = fluentbit_server->listen(2);` is called in the test setup.
`accept()` is just popping connections off a queue.  

But I agree.  It would be more readable to move these up.

Reply:
Oh, turns out there is a race here, when using the poll socket.  

When you call `accept()` on a poll socket, we poll the socket for a read event.  Calling this twice in a row may cause the polls to conflict with each other, since they are polling the same socket.  So when I moved the `accept()`s up, the second one would fail with:
`(_client2).failure(): Failed to accept: Resource temporarily unavailable`

The reason why the `accept()`s passed when placed later is because two sockets were ready to be accepted.  When placed early, no sockets are ready.

Reply:
A non-racy way to capture these sockets would be:
```
  // We expect two connections to be made with our `fluentbit_server`:
  // one for stdout and one for stderr.
  Future<network::inet::Socket> _client1 = fluentbit_server->accept();
  AWAIT_ASSERT_READY(_client1);
  
  Future<network::inet::Socket> _client2 = fluentbit_server->accept();
  AWAIT_ASSERT_READY(_client2);
```
Placed after the task has entered the STARTING phase.